### ISSUE TITLE
TF 2.4.1 yields NaNs on Large Models using model.fit and Adam as compared to 2.3.0

### ISSUE THREAD

**Issue by nicolasshu** [tensorflow/tensorflow#47649]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 2.4.1
- Python version: 3.7

**Describe the current behavior**
Whenever I train this model on TF 2.4.1, after the second batch, it starts to output NaNs as the loss. However, whenever I run this same exact code on TF 2.3.0, it trains properly. The Adam optimizer has the same hyperparameters in both versions. I'm not sure what would cause the difference. 

**Code to Reproduce NaNs**

```python
import numpy as np
import tensorflow as tf

class MyLayer(tf.keras.layers.Layer):
    def __init__(
        self,
        input_dim: int = None,
        output_dim: int = 512,
        context_dim: int = 5,
        stride: int = 1,
        dilation: int = 1,
        kernel_initializer='glorot_uniform'
    ):
        super(MyLayer, self).__init__()
        self.input_dim = input_dim
        self.conv = tf.keras.layers.Conv1D(filters=output_dim,
                                           kernel_size=context_dim,
                                           strides=stride,
                                           dilation_rate=dilation,
                                           kernel_initializer=kernel_initializer)

    def call(self, x):
        batch_size, num_frames, num_feats = x.shape
        # if self.input_dim:
        #     assert self.input_dim == num_feats
        return self.conv(x)

class Reg(tf.keras.layers.Layer):
    def __init__(self,dropout_rate=0.2):
        super(Reg,self).__init__()
        self.bn = tf.keras.layers.BatchNormalization()
        self.do = tf.keras.layers.Dropout(rate=dropout_rate)
    def call(self,x,training=True):
        x = self.bn(x,training=training)
        return self.do(x,training=training)
        
class MyModel(tf.keras.Model):
    def __init__(self, input_dim, output_dim,dropout_rate=0.2,batch_norm=False, return_xvector=False):
        super().__init__()

        self.input_dim = input_dim
        self.output_dim = output_dim

        self.fc1 = MyLayer(input_dim=self.input_dim,output_dim=512,context_dim=5,dilation=1)
        self.fc2 = MyLayer(input_dim=512  ,output_dim=1536 ,context_dim=3, dilation=2)
        self.fc3 = MyLayer(input_dim=1536 ,output_dim=512  ,context_dim=3, dilation=3)
        self.fc4 = MyLayer(input_dim=512  ,output_dim=512  ,context_dim=1, dilation=1)
        self.fc5 = MyLayer(input_dim=512  ,output_dim=1500 ,context_dim=1, dilation=1)

        self.fc6 = tf.keras.layers.Dense(512)
        self.fc7 = tf.keras.layers.Dense(512)
        self.output_layer = tf.keras.layers.Dense(self.output_dim)

        if batch_norm:
            self.reg1 = Reg(dropout_rate)
            self.reg2 = Reg(dropout_rate)
            self.reg3 = Reg(dropout_rate)
            self.reg4 = Reg(dropout_rate)
            self.reg5 = Reg(dropout_rate)
            self.reg6 = Reg(dropout_rate)
        else:
            self.reg1 = tf.keras.layers.Dropout(rate=dropout_rate)
            self.reg2 = tf.keras.layers.Dropout(rate=dropout_rate)
            self.reg3 = tf.keras.layers.Dropout(rate=dropout_rate)
            self.reg4 = tf.keras.layers.Dropout(rate=dropout_rate)
            self.reg5 = tf.keras.layers.Dropout(rate=dropout_rate)
            self.reg6 = tf.keras.layers.Dropout(rate=dropout_rate)



    def call(self,x,training=True, return_logits=True):
        with tf.name_scope("Extractor"):
            with tf.name_scope("Fc1"):
                x = tf.nn.relu(self.fc1(x))
                x = self.reg1(x, training=training)
            with tf.name_scope("Fc2"):
                x = tf.nn.relu(self.fc2(x))
                x = self.reg2(x, training=training)
            with tf.name_scope("Fc3"):
                x = tf.nn.relu(self.fc3(x))
                x = self.reg3(x, training=training)
            with tf.name_scope("Fc4"):
                x = tf.nn.relu(self.fc4(x))
                x = self.reg4(x, training=training)
            with tf.name_scope("Fc5"):
                x = tf.nn.relu(self.fc5(x))
                x = self.reg5(x, training=training)
            with tf.name_scope("StatsPool"):
                x = self.statpool(x)

            with tf.name_scope("Segment6"):
                x = self.fc6(x)
        
        with tf.name_scope("Classifier"):
            x = tf.nn.relu(x)
            x = self.reg6(x)
            x = tf.nn.relu(self.fc7(x))
            x = self.output_layer(x)
        if return_logits:
            return x 
        else:
            x = self.softmax(x)
            return x

    def statpool(self,x):
        mu = tf.math.reduce_mean(x,axis=1)
        sigma = tf.math.reduce_std(x,axis=1)
        return tf.concat([mu,sigma],1)
        
n_feats = 40
model = MyModel(n_feats,3)

loss_fn =  tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
model.compile(loss=loss_fn, optimizer="adam",metrics = ["accuracy"])

def batch(N=4):
    for i in range(100):
        yield np.random.normal(size=(N,n_feats,200)), np.random.randint(0,3,size=(N))

history = model.fit(x = batch(N=300),
                    validation_data= batch(N=300),
                   )
```

**Comment by Saduf2019** [comment]
@nicolasshu 
This is models issue, can you please open it in models repo.

**Comment by nicolasshu** [comment]
They asked me to keep the issue here
[https://github.com/tensorflow/models/issues/9802#issuecomment-800798183](https://github.com/tensorflow/models/issues/9802#issuecomment-800798183)

**Comment by Saduf2019** [comment]
@nicolasshu 
I ran the code shared and face a [different error](https://colab.research.google.com/gist/Saduf2019/0d59c4e8588568666eedebe74117a05e/untitled563.ipynb) than reported, please share all dependencies for us to replicate the issue.

**Comment by nicolasshu** [comment]
Ah I apologize. The first line was not supposed to be included:

```python
import numpy as np
import tensorflow as tf

class MyLayer(tf.keras.layers.Layer):
    def __init__(
        self,
        input_dim: int = None,
        output_dim: int = 512,
        context_dim: int = 5,
        stride: int = 1,
        dilation: int = 1,
        kernel_initializer='glorot_uniform'
    ):
        super(MyLayer, self).__init__()
        self.input_dim = input_dim
        self.conv = tf.keras.layers.Conv1D(filters=output_dim,
                                           kernel_size=context_dim,
                                           strides=stride,
                                           dilation_rate=dilation,
                                           kernel_initializer=kernel_initializer)

    def call(self, x):
        batch_size, num_frames, num_feats = x.shape
        # if self.input_dim:
        #     assert self.input_dim == num_feats
        return self.conv(x)

class Reg(tf.keras.layers.Layer):
    def __init__(self,dropout_rate=0.2):
        super(Reg,self).__init__()
        self.bn = tf.keras.layers.BatchNormalization()
        self.do = tf.keras.layers.Dropout(rate=dropout_rate)
    def call(self,x,training=True):
        x = self.bn(x,training=training)
        return self.do(x,training=training)
        
class MyModel(tf.keras.Model):
    def __init__(self, input_dim, output_dim,dropout_rate=0.2,batch_norm=False, return_xvector=False):
        super().__init__()

        self.input_dim = input_dim
        self.output_dim = output_dim

        self.fc1 = MyLayer(input_dim=self.input_dim,output_dim=512,context_dim=5,dilation=1)
        self.fc2 = MyLayer(input_dim=512  ,output_dim=1536 ,context_dim=3, dilation=2)
        self.fc3 = MyLayer(input_dim=1536 ,output_dim=512  ,context_dim=3, dilation=3)
        self.fc4 = MyLayer(input_dim=512  ,output_dim=512  ,context_dim=1, dilation=1)
        self.fc5 = MyLayer(input_dim=512  ,output_dim=1500 ,context_dim=1, dilation=1)

        self.fc6 = tf.keras.layers.Dense(512)
        self.fc7 = tf.keras.layers.Dense(512)
        self.output_layer = tf.keras.layers.Dense(self.output_dim)

        if batch_norm:
            self.reg1 = Reg(dropout_rate)
            self.reg2 = Reg(dropout_rate)
            self.reg3 = Reg(dropout_rate)
            self.reg4 = Reg(dropout_rate)
            self.reg5 = Reg(dropout_rate)
            self.reg6 = Reg(dropout_rate)
        else:
            self.reg1 = tf.keras.layers.Dropout(rate=dropout_rate)
            self.reg2 = tf.keras.layers.Dropout(rate=dropout_rate)
            self.reg3 = tf.keras.layers.Dropout(rate=dropout_rate)
            self.reg4 = tf.keras.layers.Dropout(rate=dropout_rate)
            self.reg5 = tf.keras.layers.Dropout(rate=dropout_rate)
            self.reg6 = tf.keras.layers.Dropout(rate=dropout_rate)



    def call(self,x,training=True, return_logits=True):
        with tf.name_scope("Extractor"):
            with tf.name_scope("Fc1"):
                x = tf.nn.relu(self.fc1(x))
                x = self.reg1(x, training=training)
            with tf.name_scope("Fc2"):
                x = tf.nn.relu(self.fc2(x))
                x = self.reg2(x, training=training)
            with tf.name_scope("Fc3"):
                x = tf.nn.relu(self.fc3(x))
                x = self.reg3(x, training=training)
            with tf.name_scope("Fc4"):
                x = tf.nn.relu(self.fc4(x))
                x = self.reg4(x, training=training)
            with tf.name_scope("Fc5"):
                x = tf.nn.relu(self.fc5(x))
                x = self.reg5(x, training=training)
            with tf.name_scope("StatsPool"):
                x = self.statpool(x)

            with tf.name_scope("Segment6"):
                x = self.fc6(x)
        
        with tf.name_scope("Classifier"):
            x = tf.nn.relu(x)
            x = self.reg6(x)
            x = tf.nn.relu(self.fc7(x))
            x = self.output_layer(x)
        if return_logits:
            return x 
        else:
            x = self.softmax(x)
            return x

    def statpool(self,x):
        mu = tf.math.reduce_mean(x,axis=1)
        sigma = tf.math.reduce_std(x,axis=1)
        return tf.concat([mu,sigma],1)
        
n_feats = 40
model = MyModel(n_feats,3)

loss_fn =  tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
model.compile(loss=loss_fn, optimizer="adam",metrics = ["accuracy"])

def batch(N=4):
    for i in range(100):
        yield np.random.normal(size=(N,n_feats,200)), np.random.randint(0,3,size=(N))

history = model.fit(x = batch(N=300),
                    validation_data= batch(N=300),
                   )
```

**Comment by Saduf2019** [comment]
I am able to replicate the issue for[ tf 2.4 and nightly](https://colab.research.google.com/gist/Saduf2019/ca45262c891488013dda9960db673de3/untitled563.ipynb), it works as expected on tf 2.3, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/4d576cdb603e67da3947a6539484b17f/untitled564.ipynb)

**Comment by sushreebarsa** [comment]
Was able to replicate the issue in TF v2.5,please find the gist[ here](https://colab.research.google.com/gist/sushreebarsa/2c1b2a7e9055fdb8095e23b2b1bcf0ab/untitled154.ipynb)..Thanks !

**Comment by nicolasshu** [comment]
@rajaditya-m Totally not a workaround, but no one ever got back to me, so I just downgraded.

**Comment by tilakrayal** [comment]
Hi,

Could you please try the code in Keras 3.
You can install Keras 3 using pip install -U keras and import keras directly and let us know the outcome. Also could you try converting labels into categories like this
labels = keras.utils.to_categorical(labels, 8)
and try to use the loss function categorical_crossentropy

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
